satellite/satellitedb/certdb.go

```diff
@@ -19,82 +19,85 @@ type certDB struct {
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
+
+	if pi == nil {
+		return Error.New("Peer Identity cannot be nil")
+	}
+	chain := identity.EncodePeerIdentity(pi)
+
+	var id []byte
+	query := `SELECT node_id FROM peerIdentities WHERE serial_number = ?;`
+	err = tx.QueryRow(certs.db.Rebind(query), pi.Leaf.SerialNumber.Bytes()).Scan(&id)
 	if err != nil {
 		if err == sql.ErrNoRows {
-			return certs.tryAddPublicKey(ctx, nodeID, publicKey)
+			// create a new entry
+			_, err = tx.Exec(certs.db.Rebind(`INSERT INTO peerIdentities ( serial_number, peer_identity, node_id, update_at ) VALUES ( ?, ?, ?, ? );`), pi.Leaf.SerialNumber.Bytes(), chain, nodeID.Bytes(), time.Now())
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			return nil
 		}
 		return Error.Wrap(err)
 	}
 
-	// publickey for the nodeID entry already exists, just return
+	// already public key exists, just return
 	return nil
 }
 
-func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
+// Get gets the public key based on the certificate's serial number
+func (certs *certDB) Get(ctx context.Context, nodeID storj.NodeID) (_ *identity.PeerIdentity, err error) {
 	defer mon.Task()(&ctx)(&err)
-	// no rows err, so create/insert an entry
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
+	dbxInfo, err := certs.db.Get_PeerIdentity_By_NodeId_OrderBy_Desc_UpdateAt(ctx, dbx.PeerIdentity_NodeId(nodeID.Bytes()))
 	if err != nil {
-		return Error.Wrap(err)
+		return nil, Error.Wrap(err)
 	}
 
-	// TODO: use upsert here instead of create
-	_, err = certs.db.Create_CertRecord(ctx,
-		dbx.CertRecord_Publickey(pubbytes),
-		dbx.CertRecord_Id(nodeID.Bytes()),
-	)
-	// another goroutine might race to create the cert record, let's ignore that error
-	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
-		return nil
-	} else if err != nil {
-		return Error.Wrap(err)
+	if dbxInfo == nil {
+		return nil, Error.New("unknown nodeID :%+v: %+v", nodeID.Bytes(), err)
 	}
 
-	return nil
+	peer, err := identity.DecodePeerIdentity(ctx, dbxInfo.PeerIdentity)
+	return peer, Error.Wrap(err)
 }
 
-// GetPublicKey gets the public key of uplink corresponding to uplink id
-func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ crypto.PublicKey, err error) {
+// BatchGet gets the public key based on the certificate's serial number
+func (certs *certDB) BatchGet(ctx context.Context, nodeIDs []storj.NodeID) (peers []*identity.PeerIdentity, err error) {
 	defer mon.Task()(&ctx)(&err)
-	dbxInfo, err := certs.db.All_CertRecord_By_Id_OrderBy_Desc_UpdateAt(ctx, dbx.CertRecord_Id(nodeID.Bytes()))
-	if err != nil {
-		return nil, err
+	if len(nodeIDs) == 0 {
+		return nil, nil
 	}
 
-	// the first indext always holds the lastest of the keys
-	pubkey, err := pkcrypto.PublicKeyFromPKIX(dbxInfo[0].Publickey)
+	tx, err := certs.db.Open(ctx)
```

Comment:
Don't use a transaction here, there's no need for it, query all at the same time instead.

Reply:
made change to request all nodeIDs peer identities at the one shot